### DISCUSSION TITLE
post_232218

### DISCUSSION THREAD

**Discussion by cincuranet** [cincuranet/blog-comments#634] [Comments]
imported 

https://www.tabsoverspaces.com/id/232218

**Comment by cincuranet** [comment]
Imported comment written by _Thomas Steinmaurer_ on _2010-12-15T04:55:16_:

<p>Nice to see, that you adopt the Trace API stuff in the .NET provider. ;-)</p>

**Comment by cincuranet** [comment]
Imported comment written by _Gabriele_ on _2010-12-15T23:48:16_:

<p>Hi great job! Have you thought about a class to trace and audit directly to database's table.<br>In my fb2.5 I would like to audit all changes (I,U,D) made to every table in my fdb writing to audit_table the following info :<br>table name, field name, old value, new value, user, action type (Insert,Update,Delete)<br>I know I could solve it creating a single trigger for every table..is there a simpler way ?<br>What about it ? Thanx</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2010-12-16T00:11:38_:

<p>I don't think the Trace API was designed for this. Your application would need to be constantly running and parsing the result.</p>

**Comment by cincuranet** [comment]
Imported comment written by _Gabriele_ on _2010-12-16T15:56:09_:

<p>Yes, you're right.  Just an idea.You say "some methods may be added"... Well, just think about auditing changes.It would be a great improvement Thank you from Italy!</p>

**Comment by wowotec** [comment]
I have a problem with this code. Everything connects and works outside of the Start method. I have 
response from the List method and information about creating a session eg. "Trace session ID 99 started" but I don't get any reply. What to do to get responses from the ServiceOutputEventArgs every time something is registered?

**Comment by cincuranet** [comment]
Something has to happen in the database (and that something needs to be what you registered for).

**Comment by wowotec** [comment]
I have no idea why it is not working. When I use fbprofiler-1.0.5-win32 it is reciving all information from server. When using FbTrace() I have only information about session ID. When I use any other method they are work correctlly.. My code is:

```
public async Task Run()
        {

            FbTrace trace = new FbTrace(connectionString: DbTrace.GetConnectionString());

            trace.DatabasesConfigurations.Add(
                new FbDatabaseTraceConfiguration()
                {
                    DatabaseName = string.Empty,
                    Enabled = true,
                    Events = FbDatabaseTraceEvents.BLRRequests | FbDatabaseTraceEvents.Connections | FbDatabaseTraceEvents.Context
                        | FbDatabaseTraceEvents.DYNRequests | FbDatabaseTraceEvents.Errors | FbDatabaseTraceEvents.ExplainPlan
                        | FbDatabaseTraceEvents.FunctionFinish | FbDatabaseTraceEvents.FunctionStart | FbDatabaseTraceEvents.InitFini
                        | FbDatabaseTraceEvents.PrintBLR | FbDatabaseTraceEvents.PrintDYN | FbDatabaseTraceEvents.PrintPerf
                        | FbDatabaseTraceEvents.PrintPlan | FbDatabaseTraceEvents.ProcedureFinish | FbDatabaseTraceEvents.ProcedureStart
                        | FbDatabaseTraceEvents.StatementFinish | FbDatabaseTraceEvents.StatementFree | FbDatabaseTraceEvents.StatementPrepare
                        | FbDatabaseTraceEvents.StatementStart | FbDatabaseTraceEvents.Sweep | FbDatabaseTraceEvents.Transactions
                        | FbDatabaseTraceEvents.TriggerFinish | FbDatabaseTraceEvents.TriggerStart | FbDatabaseTraceEvents.Warnings
                    ,
                    IncludeFilter = "%(SELECT|INSERT|UPDATE|DELETE)%"
                    
                });

            var sessionId = -1;
            trace.ServiceOutput += (sender, e) =>
            {
                if (sessionId == -1)
                {
                    var match = Regex.Match(e.Message, @"Trace session ID (\d+) started");
                    if (match.Success)
                    {
                        sessionId = int.Parse(match.Groups[1].Value);
                    }
                }
                ServiceOutput(sender, e);
            };
            

            async Task Stop()
            {
                await Task.Delay(20000);
                await new FbTrace(connectionString: DbTrace.GetConnectionString())
                    .StopAsync(sessionId);
            }
            var stopTask = Stop();
            await trace.StartAsync("Testing",CancellationToken.None);
            await stopTask;

        }

        private static void ServiceOutput(object sender, ServiceOutputEventArgs e)
        {
            Console.WriteLine(e.Message);
        }
```

Do you have any idea what I'm doing wrgong?

**Comment by cincuranet** [comment]
Your code works fine for me:
![img](https://user-images.githubusercontent.com/4540597/131462614-16cd1442-22f5-412c-b38d-8ebda490d1cd.png)
